README.md
# Foreman Docker Plugin

[![Code Climate](https://codeclimate.com/github/theforeman/foreman-docker/badges/gpa.svg)](https://codeclimate.com/github/theforeman/foreman-docker)
[![Gem Version](https://badge.fury.io/rb/foreman_docker.svg)](http://badge.fury.io/rb/foreman_docker)
[![Dependency Status](https://gemnasium.com/theforeman/foreman-docker.svg)](https://gemnasium.com/theforeman/foreman-docker)
[![Issue Stats](http://issuestats.com/github/theforeman/foreman-docker/badge/pr)](http://issuestats.com/github/theforeman/foreman-docker)

```foreman_docker``` enables provisioning and managing of [Docker](http://docker.com) containers and images in [Foreman](http://github.com/theforeman/foreman), all of that under the GPL v3+ license.

* Website: [TheForeman.org](http://theforeman.org)
* ServerFault tag: [Foreman](http://serverfault.com/questions/tagged/foreman)
* Issues: [foreman_docker Redmine](http://projects.theforeman.org/projects/docker/issues)
* Wiki: [Foreman wiki](http://projects.theforeman.org/projects/foreman/wiki/About)
* Community and support: #theforeman for general support, #theforeman-dev for development chat in [Freenode](irc.freenode.net)
* Mailing lists:
    * [foreman-users](https://groups.google.com/forum/?fromgroups#!forum/foreman-users)
    * [foreman-dev](https://groups.google.com/forum/?fromgroups#!forum/foreman-dev)

## Features

* Special view with logs and processes of Foreman managed containers
    ![](http://i.imgur.com/D21bdgj.png)
    ![](http://i.imgur.com/XnrPTZC.png)
* Wizard for container creation and cgroups configuration
    ![Select a docker image](http://i.imgur.com/IoMuNnr.png)
    ![Cgroups configuration](http://i.imgur.com/74d99Tf.png)
* Commit and upload containers: creates an image with the status of your current container
    ![Commit and upload to the docker hub](http://i.imgur.com/coF5Y0L.png)
* Container listing and basic CRUD operations
    ![](http://i.imgur.com/DPcaHkZ.png)

### Planned
* [Kubernetes](https://github.com/kubernetes/kubernetes/) integration
* Events stream ([#8037](http://projects.theforeman.org/issues/8037))
* Tight integration between Docker hosts [Atomic](http://www.projectatomic.io/) and [CoreOS](http://coreos.com/) and containers ([#7653](http://projects.theforeman.org/issues/7653), [#7652](http://projects.theforeman.org/issues/7652))
* Quickstart images - pre-supplied images and configuration ([#7869](http://projects.theforeman.org/issues/7869))
* Links to other containers ([#7866](http://projects.theforeman.org/issues/7866))
* API ([#7874](http://projects.theforeman.org/issues/7874))
* [Hammer CLI](http://github.com/theforeman/hammer-cli-foreman) support ([#8227](http://projects.theforeman.org/issues/8227))

## Installation

Please see the Foreman manual for appropriate instructions:

* [Foreman: How to Install a Plugin](http://theforeman.org/manuals/latest/index.html#6.1InstallaPlugin)

### Red Hat, CentOS, Fedora, Scientific Linux (rpm)

Set up the repo as explained in the link above, then run

    # yum install ruby193-rubygem-foreman_docker

### Debian, Ubuntu (deb)

Set up the repo as explained in the link above, then run

    # apt-get install ruby-foreman-docker

### Bundle (gem)

Add the following to bundler.d/Gemfile.local.rb in your Foreman installation directory (/usr/share/foreman by default)

    $ gem 'foreman_docker'

Then run `bundle install` and `foreman-rake db:migrate` from the same directory

--------------

To verify that the installation was successful, go to Foreman, top bar **Administer > About** and check 'foreman_docker' shows up in the **System Status** menu under the Plugins tab. You should also see a **'Containers'** button show up in the top bar, similar to this

![](http://i.imgur.com/Ug14Ktl.png)

## Configuration

Go to **Infrastructure > Compute Resources** and click on "New Compute Resource".

Choose the **Docker provider**, and fill in all the fields. User name, password, and email are used so that Docker clients such as Foreman can make the host download images from the Docker hub. Your password will be encrypted in the database.

That's it. You're now ready to create and manage containers in your new Docker compute resource.

## Compatibility

| Foreman | Plugin |
| ---------------:| --------------:|
| >= 1.5         | 0.0.1 - 0.0.3   |
| >= 1.6         | 0.1.0 - 0.2.0   |
| >= 1.7         | 1.0.0 - 2.1.1   |
| >= 1.7         | 3.0.0+          |
| >= 1.15        | 3.1.0+          |

## Docker Registry API Compatibility

| Plugin version | Registry API version |
| ---------------:| --------------:|
| <  3.1.0        | [v1](http://docs.master.dockerproject.org/v1.7/reference/api/registry_api/)             |
| >= 3.1.0        | [v1](http://docs.master.dockerproject.org/v1.7/reference/api/registry_api/), [v2](https://docs.docker.com/registry/spec/api/)*****    |

_*** Note:** API v2 as default and v1 to fall back on._

See extras/RELEASE.md for more detailed information on compatibility and releases.

## How to contribute?

Generally, follow the [Foreman guidelines](http://theforeman.org/contribute.html). For code-related contributions, fork this project and send a pull request with all changes. Some things to keep in mind:
* Code from the master branch can contain features only present in [Fog's](http://github.com/fog/fog) master branch, we commit to wait for the next Fog release to put that code in a foreman-docker release.
* [Follow the rules](http://theforeman.org/contribute.html#SubmitPatches) about commit message style and create a Redmine issue. Doing this right will help reviewers to get your contribution merged faster.
* [Rubocop](https://github.com/bbatsov/rubocop) will analyze your code, you can run it locally with `rake rubocop`.
* All of our pull requests run the full test suite in our [Jenkins CI system](http://ci.theforeman.org/). Please include tests in your pull requests for any additions or changes in functionality


### Testing

Run `rake test:docker` from your Foreman directory to run the test suite.

## Latest code

You can get the develop branch of the plugin by specifying your Gemfile in this way:

    gem 'foreman_docker', :git => "https://github.com/theforeman/foreman-docker.git"

# Copyright

Copyright (c) 2014 Amos Benari

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG
2019-02-26	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Merge pull request #216 from ares/fix/19591

2019-02-22	Marek Hulan <mhulan@redhat.com>
	* Fixes #19591 - provide clean up task

2018-04-16	Daniel Lobato Garcia <me@daniellobato.me>
	* Release 4.1.0

2018-04-16	Stephen Benjamin <stephen@bitbin.de>
	* refs #23019 - don't pass class to class_name in assocs (#211)

2018-01-11	Daniel Lobato García <me@daniellobato.me>
	* Fixes #20067 - Wrong constant name Docker/ImageSearch

2018-01-10	Daniel Lobato García <me@daniellobato.me>
	* Fix test redirect_to_back

2018-01-09	Daniel Lobato Garcia <me@daniellobato.me>
	* Remove redirect_to :back (Rails 5.1)

2018-01-09	Michael Moll <kvedulv@kvedulv.de>
	* More Rails 5.1 changes

2018-01-05	Tim Meusel <tim@bastelfreak.de>
	* update link to kubernetes github repo

2017-12-21	Daniel Lobato Garcia <me@daniellobato.me>
	* Version 4.0.0

2017-12-21	Michael Moll <kvedulv@kvedulv.de>
	* Update to Rails 5.1

2017-12-14	Daniel Lobato Garcia <me@daniellobato.me>
	* Refs #21893 - rubocop trailing whitespace

2017-12-14	Marek Hulán <ares@users.noreply.github.com>
	* Fixes #21893 - fix listing containers for specific CR

2017-12-13	Swapnil Abnave <sabnave@redhat.com>
	* Fixes #20429 - mandate URL for docker compute-resource

2017-12-13	Michael Moll <kvedulv@kvedulv.de>
	* Fixes #21921 - Upgrade rubocop to 0.51

2017-11-24	Daniel Lobato Garcia <me@daniellobato.me>
	* FactoryGirl to FactoryBot

2017-11-22	Ondřej Pražák <xprazak2@users.noreply.github.com>
	* Fixes #21732 - Replace :before_filter with :before_action

2017-10-06	Walden Raines <walden@redhat.com>
	* Fixes #21143: add containers icon for vertical navigation.

2017-07-31	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #20355 - Request registry results in JSON

2017-06-28	Sebastian Gräßl <mail@bastilian.me>
	* Bump version to 3.2.1
	* Revert "Fixes #19964 - Replace ActiveRecord

2017-06-27	Sebastian Gräßl <mail@bastilian.me>
	* Bump version to 3.2.0

2017-06-27	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #19964 - Replace ActiveRecord

2017-06-23	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #19957 - Add Registry compatiblity info to README

2017-06-20	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Merge pull request #187 from bastilian/fix/18902-PR

2017-06-19	Sebastian Gräßl <mail@bastilian.me>
	* Fixes #18902 - Enable autocomplete for container image search

2017-06-19	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #16235 - Add verify SSL option for Registries

2017-06-13	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Refs #19694 - Bump rubocop version to 0.49.1

2017-06-07	Daniel Lobato Garcia <me@daniellobato.me>
	* Add compatibility info for latest version
	* Version 3.1.0

2017-06-01	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Merge pull request #192 from ares/fix/19475

2017-05-30	Marek Hulan <mhulan@redhat.com>
	* Fixes #19633 - use standard plugin test setup

2017-05-30	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #19694 - Align rubocop configuration with foreman-core

2017-05-03	Sebastian Gräßl <mail@bastilian.me>
	* Fixes #19396 - Rename test variable name to more specific term

2017-04-06	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Merge pull request #186 from bastilian/fix/18518-PR2

2017-04-06	Marek Hulán <ares@users.noreply.github.com>
	* Fixes #17960 - add plugin permissions to core

2017-03-23	Sebastian Gräßl <mail@bastilian.me>
	* Fixes #18518 - Add proper validation for container image

2017-03-09	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #18733 - Prevent fallback to Docker Hub on registry search
	* Fixes #14175 - Add support for Docker Registry v2

2017-02-15	Sebastian Gräßl <bastilian@users.noreply.github.com>
	* Fixes #18502 - Add excon version 0.46 and above as dependency

2017-01-13	June Zhang <junzhang@redhat.com>
	* Fixes #18021 - Add missing files for updating New buttons

2017-01-11	June Zhang <junzhang@redhat.com>
	* Fixes #18021 - update the button label from New to Create

2016-12-20	Swapnil Abnave <sabnave@redhat.com>
	* Fixes #17729 - APIdoc - URL desc add 'Docker' to CR types

2016-11-17	orrabin <orabin@redhat.com>
	* Refs #14370 - Adding new_link helper to containers and registries

2016-09-26	orrabin <orabin@redhat.com>
	* Fixes #13043 - Change Parameter inheritance to DockerParameter

2016-09-26	Daniel Lobato García <me@daniellobato.me>
	* Refs #9752 - Test Docker errors are propagated

2016-09-20	Ranjan Kumar <ranjankumar188@gmail.com>
	* Fixes #16090 - Docker container wizard missing image false.png

2016-09-19	June Zhang <junzhang@redhat.com>
	* Fixes #16594 - Capitalize the first letter of every word

2016-08-01	Daniel Lobato <elobatocs@gmail.com>
	* Version 3.0.0

2016-08-01	David Davis <daviddavis@redhat.com>
	* Fixes #15888 - Remove attr_accessible calls

2016-06-28	Shlomi Zadok <shlomi@ben-hanna.com>
	* Fixes #15528 - 'command' should not be required

2016-06-23	Shlomi Zadok <shlomi@ben-hanna.com>
	* Refs #14570 - Destroy Container db object When deleting with container uuid params, the container object should be destroyed as well

2016-05-30	Daniel Lobato <elobatocs@gmail.com>
	* Refs #14570 - Fetch tags from local images properly
	* Fixes #9752 - Pass Docker errors to controller
	* Fixes #15135 - Fix JS error in datatables

2016-05-27	Daniel Lobato <elobatocs@gmail.com>
	* Fixes #15098 - Scope registry by taxonomies
	* Ensure Fog is mocked before running API/v2/create test
	* Use remove_reference to remove FK with Rails 4

2016-05-20	Daniel Lobato <elobatocs@gmail.com>
	* Fixes #15107 - Rails 4.2 test breakages

2016-04-19	Daniel Lobato <elobatocs@gmail.com>
	* Fixes #14703 - Add taxonomies to Registries API

2016-04-15	Daniel Lobato <elobatocs@gmail.com>
	* Release 2.1.1
	* Fixes #14662 - Revert new_link to ensure 1.11 compatibility
	* Release 2.1.0
	* Fixes #9270 - Add email validation

lib/foreman_docker/engine.rb
require 'fast_gettext'
require 'gettext_i18n_rails'
require 'fog/fogdocker'
require 'wicked'
require 'docker'
require 'deface'

module ForemanDocker
  # Inherit from the Rails module of the parent app (Foreman), not the plugin.
  # Thus, inherits from ::Rails::Engine and not from Rails::Engine
  class Engine < ::Rails::Engine
    engine_name 'foreman_docker'

    config.autoload_paths += Dir["#{config.root}/app/controllers/concerns"]
    config.autoload_paths += Dir["#{config.root}/app/models/concerns"]

    initializer 'foreman_docker.load_app_instance_data' do |app|
      ForemanDocker::Engine.paths['db/migrate'].existent.each do |path|
        app.config.paths['db/migrate'] << path
      end
    end

    foreman_docker_assets = %w[foreman_docker/autocomplete.css
                               foreman_docker/terminal.css
                               foreman_docker/container_image_search.js
                               foreman_docker/create_registry.js]

    initializer "foreman_docker.assets.precompile" do |app|
      app.config.assets.precompile += foreman_docker_assets
    end

    initializer 'foreman_docker.configure_assets', :group => :assets do
      SETTINGS[:foreman_docker] =
        { :assets => { :precompile => foreman_docker_assets } }
    end

    initializer 'foreman_docker.register_gettext', :after => :load_config_initializers do
      locale_dir = File.join(File.expand_path('../../..', __FILE__), 'locale')
      locale_domain = 'foreman_docker'

      Foreman::Gettext::Support.add_text_domain locale_domain, locale_dir
    end

    initializer 'foreman_docker.register_plugin', :before => :finisher_hook do
      Foreman::Plugin.register :foreman_docker do
        requires_foreman '>= 1.15'
        compute_resource ForemanDocker::Docker

        sub_menu :top_menu, :containers_menu, :caption => N_('Containers'),
                                              :icon => 'fa fa-cube',
                                              :after => :monitor_menu do
          menu :top_menu, :containers,    :caption => N_('All Containers'),
                                          :url_hash => { :controller => :containers,
                                                         :action => :index }
          menu :top_menu, :new_container, :caption => N_('Create Container'),
                                          :url_hash => { :controller => :containers,
                                                         :action => :new }
          menu :top_menu, :registries, :caption => N_('Registries'),
                                       :url_hash => { :controller => :registries,
                                                      :action => :index }
        end

        security_block :containers do
          permission :view_containers,
                     { :containers          => [:index, :show],
                       :'api/v2/containers' => [:index, :show, :logs] },
                     :resource_type => 'Container'
          permission :commit_containers, { :containers => [:commit] },
                     :resource_type => 'Container'
          permission :create_containers,
                     { :'containers/steps'  => [:show, :update],
                       :containers          => [:new],
                       :'api/v2/containers' => [:create, :power] },
                     :resource_type => 'Container'
          permission :destroy_containers,
                     { :containers          => [:destroy],
                       :'api/v2/containers' => [:destroy] },
                     :resource_type => 'Container'
          permission :power_compute_resources_vms,
                     { :containers          => [:power],
                       :'api/v2/containers' => [:create, :power] },
                     :resource_type => 'ComputeResource'
        end

        security_block :registries do
          permission :view_registries,
                     { :registries => [:index, :show],
                       :'api/v2/registries' => [:index, :show] },
                     :resource_type => 'DockerRegistry'
          permission :create_registries,
                     { :registries  => [:new, :create, :update, :edit],
                       :'api/v2/registries' => [:create, :update] },
                     :resource_type => 'DockerRegistry'
          permission :destroy_registries,
                     { :registries => [:destroy],
                       :'api/v2/registries' => [:destroy] },
                     :resource_type => 'DockerRegistry'
        end

        security_block :image_search do
          permission :search_repository_image_search,
                     { :image_search => [:auto_complete_repository_name,
                                         :auto_complete_image_tag,
                                         :search_repository] },
                     :resource_type => 'DockerRegistry'
        end

        add_all_permissions_to_default_roles

        parameter_filter ComputeResource, :email

        # apipie API documentation
        # Only available in 1.8, otherwise it has to be in the initializer below
        if SETTINGS[:version].to_s.include?('develop') ||
           Gem::Version.new(SETTINGS[:version].notag) >= Gem::Version.new('1.8')
          apipie_documented_controllers [
            "#{ForemanDocker::Engine.root}/app/controllers/api/v2/*.rb"]
        end
      end
    end

    initializer "foreman_docker.apipie" do
      # this condition is here for compatibility reason to work with Foreman 1.4.x
      # Also need to handle the reverse of the 1.8 method above
      unless SETTINGS[:version].to_s.include?('develop') ||
             Gem::Version.new(SETTINGS[:version].notag) >= Gem::Version.new('1.8')
        if Apipie.configuration.api_controllers_matcher.is_a?(Array)
          Apipie.configuration.api_controllers_matcher <<
            "#{ForemanDocker::Engine.root}/app/controllers/api/v2/*.rb"
        end
      end
    end

    require 'fog/fogdocker/models/compute/server'
    require 'fog/fogdocker/models/compute/image'
    require 'fog/fogdocker/models/compute/images'
    require File.expand_path('../../../app/models/concerns/fog_extensions/fogdocker/server',
                             __FILE__)
    require File.expand_path('../../../app/models/concerns/fog_extensions/fogdocker/image',
                             __FILE__)
    require File.expand_path('../../../app/models/concerns/fog_extensions/fogdocker/images',
                             __FILE__)
    config.to_prepare do
      Fog::Compute::Fogdocker::Server.send(:include, ::FogExtensions::Fogdocker::Server)
      Fog::Compute::Fogdocker::Image.send(:include, ::FogExtensions::Fogdocker::Image)
      # Compatibility fixes - to be removed once 1.7 compatibility is no longer required
      Fog::Compute::Fogdocker::Images.send(:include, ::FogExtensions::Fogdocker::Images)
      ::Taxonomy.send(:include, ForemanDocker::TaxonomyExtensions)
      ComputeResource.send(:include, ForemanDocker::ComputeResourceExtensions)
    end
  end
end
